gramineproject/gramine
# [LibOS,PAL] Emulate file-backed mmap via PAL read/write APIs

## Description of the changes 

Previously, the `chroot` FS (plain host-backed files) used the `PalStreamMap()` PAL API for file-backed mmap. This had three problems: 1) need to implement a non-trivial `map` callback in PALs; 2) discrepancy between `map` implementations in different PALs; 3) hard to debug file-mmap bugs because they only reproduced on SGX (`gramine-sgx`) PAL and not on Linux (`gramine-direct`) PAL.

Note that other FSes already used emulated file-backed mmap: `tmpfs` and `encrypted` emulate such mmaps via `PalStreamRead()` and `PalStreamWrite()`.

This PR switches `chroot` FS to use emulated file-backed mmap. This way, `chroot` becomes similar in implementation to `tmpfs` and `encrypted` FSes. Only `shm` FS still uses `PalStreamMap()` because devices with shared memory have non-standard semantics of mmaps. Corresponding `file_map()` functions in PAL are removed.

This switch to emulated mmap uncovered several bugs:
- Underlying file may be shorter than the requested mmap size. In this case access beyond the last file-backed page must cause SIGBUS. Previously this semantics worked only on `gramine-direct` and wasn't implemented on `gramine-sgx` (with EDMM).
- As a consequence of the semantics above, file-growing `ftruncate()` on already-mmapped file must make newly extended file contents accessible. Previously it didn't work on `gramine-sgx` (with EDMM), now it is resolved via `prot_refresh_mmaped_from_file_handle()` call.
- `msync()` must update file contents with the mmapped-in-process contents, but only those parts that do not exceed the file size. Previously there was a bug that msync'ed even the exceeding parts.
- Applications expect `msync(MS_ASYNC)` to update file contents before the next app access to the file. Gramine instead ignored such requests, leading to accessing stale contents. We fix this bug by treating `MS_ASYNC` the same way as `MS_SYNC`. This bug was detected on LTP test `msync01`.

Several FS tests are enabled on SGX now. Generally, `gramine-sgx` now supports shared file-backed mappings, i.e. `mmap(MAP_SHARED)`.

Extracted from https://github.com/gramineproject/gramine/pull/1812.

## How to test this PR? 

Several new tests are enabled. Maybe some new LTP tests can be enabled (didn't check).


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/1818)
